spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -0,0 +1,517 @@
+
+        TokenAmount[] memory redemptions = new TokenAmount[](assetCount);
+
+        // burn notes
+        _burn(_msgSender(), notes);
+
+        // aggregating vault assets to be redeemed
+        redemptions[0].token = underlying;
+        for (uint256 i = 0; i < deployedCount_; i++) {
+            redemptions[i + 1].token = deployedAt(i);
+        }
+        for (uint256 i = 0; i < earnedCount_; i++) {
+            redemptions[i + 1 + deployedCount_].token = earnedAt(i);
+        }
+
+        // transferring assets out proportionally
+        for (uint256 i = 0; i < assetCount; i++) {
+            redemptions[i].amount = _calculateAssetShare(redemptions[i].token, notes, totalNotes);
+            redemptions[i].token.safeTransfer(_msgSender(), redemptions[i].amount);
+            _syncAsset(redemptions[i].token);
+        }
+
+        return redemptions;
+    }
+
+    /// @notice Total value of assets currently held by the vault denominated in the underlying asset.
+    function getTVL() public returns (uint256) {
+        uint256 totalAssets = 0;
+
+        // The underlying balance
+        totalAssets += underlying.balanceOf(address(this));
+
+        // The deployed asset value denominated in the underlying
+        for (uint256 i = 0; i < _deployed.length(); i++) {
+            ITranche tranche = deployedAt(i);
+            uint256 balance = tranche.balanceOf(address(this));
+            if (balance > 0) {
+                (uint256 collateralBalance, uint256 debt) = tranche.getTrancheCollateralization();
+                totalAssets += balance.mulDiv(collateralBalance, debt);
+            }
+        }
+
+        // The earned asset (perp token) value denominated in the underlying
+        for (uint256 i = 0; i < earnedCount(); i++) {
+            IERC20Upgradeable perp_ = earnedAt(i);
+            uint256 balance = perp_.balanceOf(address(this));
+            if (balance > 0) {
+                // The "earned" asset is assumed to be the perp token.
+                totalAssets += balance.mulDiv(IPerpetualTranche(address(perp_)).getAvgPrice(), PERP_UNIT_PRICE);
+            }
+        }
+
+        return totalAssets;
+    }
+
+    //--------------------------------------------------------------------------
+    // External & Public read methods
+
+    /// @notice Fetches the vault's asset token balance.
+    /// @param token The address of the asset ERC-20 token held by the vault.
+    /// @return The token balance.
+    function assetBalance(IERC20Upgradeable token) external view returns (uint256) {
+        return isVaultAsset(token) ? token.balanceOf(address(this)) : 0;
+    }
+
+    /// @notice Checks if the given token is held by the vault.
+    /// @param token The address of a token to check.
+    function isVaultAsset(IERC20Upgradeable token) public view returns (bool) {
+        return token == underlying || _deployed.contains(address(token)) || address(perp) == address(token);
```

Comment:
This is an instance where it assumes there's one earned token, but other parts of the code doesn't. I think let's just assume there's one in rest of code too.

Reply:
Yea agree, lets assume one earned token for now everywhere